File: packages/browsergrad-grad/src/python/torch_compat.py
"""browsergrad_grad.torch_compat — install a torch-namespace shim.

The shim is split across three private modules reflecting the Pile A/B/C
taxonomy from ARCHITECTURE.md / PROGRESS.md:

* `_torch_compat_real` — Pile A: real op aliases.
* `_torch_compat_limited` — Pile B: shims with explicit caveats.
* `_torch_compat_impossible` — Pile C: loud NotImplementedError stubs.

`install_torch_alias` orchestrates: builds an empty torch module, calls each
pile's install function in order, asserts the cross-pile identity invariant,
and registers everything on `sys.modules`.
"""

import sys as _bg_torch_sys
import types as _bg_torch_types

from ._torch_compat_real import install_real
from ._torch_compat_limited import install_limited
from ._torch_compat_impossible import install_impossible


def install_torch_alias():
    """Register torch / torch.nn / torch.nn.functional / torch.optim in
    sys.modules so PyTorch user code runs against browsergrad_grad.

    Calling repeatedly is safe — the alias is re-installed each time, so
    if a user has reset sys.modules between calls it'll rebuild.
    """
    import browsergrad_grad as _bg

    torch_mod = _bg_torch_types.ModuleType("torch")
    torch_mod.__doc__ = "browsergrad_grad shim under the torch namespace"

    install_real(torch_mod, _bg, _bg_torch_types)

    # Identity invariant: Pile B patches `_bg.nn.Module.to` in place, and we
    # need that patch to be visible via `torch.nn.Module.to`. That only works
    # because Pile A's `setattr(torch_nn, name, getattr(_bg.nn, name))` copies
    # the CLASS REFERENCE, not a fresh copy, so torch_nn.Module IS _bg.nn.Module.
    # If anyone ever switches Pile A to a deep copy, this assertion fires and
    # Pile B's _module_to_shim patch must move into nn.py as a real method.
    assert torch_mod.nn.Module is _bg.nn.Module, (
        "torch_compat invariant violated: torch_mod.nn.Module must be identical "
        "to _bg.nn.Module (shallow class reference). If you changed install_real "
        "to deep-copy _bg.nn, move Module.to into nn.py instead of monkey-patching."
    )

    install_limited(torch_mod, _bg, _bg_torch_types)
    install_impossible(torch_mod, _bg_torch_types)

    # Register sys.modules entries — read sub-modules off torch_mod so we
    # don't have to know whether they were added by Pile A or Pile B/C.
    _bg_torch_sys.modules["torch"] = torch_mod
    _bg_torch_sys.modules["torch.nn"] = torch_mod.nn
    _bg_torch_sys.modules["torch.nn.functional"] = torch_mod.nn.functional
    _bg_torch_sys.modules["torch.nn.utils"] = torch_mod.nn.utils
    _bg_torch_sys.modules["torch.optim"] = torch_mod.optim
    _bg_torch_sys.modules["torch.utils"] = torch_mod.utils
    _bg_torch_sys.modules["torch.utils.data"] = torch_mod.utils.data
    _bg_torch_sys.modules["torch.amp"] = torch_mod.amp
    _bg_torch_sys.modules["torch.linalg"] = torch_mod.linalg
    _bg_torch_sys.modules["torch.fx"] = torch_mod.fx
    _bg_torch_sys.modules["torch.jit"] = torch_mod.jit
    _bg_torch_sys.modules["torch.cuda"] = torch_mod.cuda
    _bg_torch_sys.modules["torch.distributed"] = torch_mod.distributed
    _bg_torch_sys.modules["torch.onnx"] = torch_mod.onnx
    _bg_torch_sys.modules["torch.quantization"] = torch_mod.quantization

    return torch_mod

File: packages/browsergrad-grad/src/python/torch_compat.ts
/**
 * `install_torch_alias()` — installs a `torch` namespace shim into
 * sys.modules so vanilla PyTorch user code runs unmodified against
 * browsergrad_grad.
 *
 * Mapping
 *   torch                       → browsergrad_grad
 *   torch.Tensor / torch.tensor → browsergrad_grad.Tensor
 *   torch.zeros / ones / randn  → browsergrad_grad equivalents
 *   torch.cat / stack           → browsergrad_grad equivalents
 *   torch.no_grad               → browsergrad_grad.no_grad
 *   torch.nn                    → browsergrad_grad.nn (copied into a new
 *                                 module so we can override .functional)
 *   torch.nn.functional         → browsergrad_grad.functional + PyTorch-name
 *                                 aliases (cross_entropy → cross_entropy_loss)
 *   torch.optim                 → browsergrad_grad.optim
 *
 * Limitations
 *   Only the subset of torch's API that browsergrad_grad implements is
 *   available. Anything else raises AttributeError. Notably absent:
 *   torch.cuda.*, torch.compile, torch.fx, torch.jit, dtype objects
 *   beyond the f32 we use internally.
 */
export const TORCH_COMPAT_PY = `
"""browsergrad_grad.torch_compat — install a torch-namespace shim."""

import sys as _bg_torch_sys
import types as _bg_torch_types


def install_torch_alias():
    """Register torch / torch.nn / torch.nn.functional / torch.optim in
    sys.modules so PyTorch user code runs against browsergrad_grad.

    Calling repeatedly is safe — the alias is re-installed each time, so
    if a user has reset sys.modules between calls it'll rebuild.
    """
    import browsergrad_grad as _bg

    torch_mod = _bg_torch_types.ModuleType("torch")
    torch_mod.__doc__ = "browsergrad_grad shim under the torch namespace"

    # Core tensor + constructors
    torch_mod.Tensor = _bg.Tensor
    torch_mod.tensor = _bg.Tensor          # torch.tensor(...) is a function
    torch_mod.zeros = _bg.zeros
    torch_mod.ones = _bg.ones
    torch_mod.randn = _bg.randn
    torch_mod.cat = _bg.cat
    torch_mod.stack = _bg.stack
    torch_mod.no_grad = _bg.no_grad

    # Numpy interop + reproducibility
    torch_mod.from_numpy = _bg.from_numpy
    torch_mod.manual_seed = _bg.manual_seed

    # Top-level math functions (PyTorch-style alternatives to methods/ops)
    torch_mod.matmul = _bg.matmul
    torch_mod.mm = _bg.mm
    torch_mod.bmm = _bg.bmm
    torch_mod.exp = _bg.exp
    torch_mod.log = _bg.log
    torch_mod.sum = _bg.sum
    torch_mod.mean = _bg.mean
    torch_mod.argmax = _bg.argmax
    torch_mod.einsum = _bg.einsum

    # Serialization
    torch_mod.save = _bg.save
    torch_mod.load = _bg.load

    # dtype tokens — strings are enough for the dtype= kwarg path we support
    torch_mod.float32 = "float32"
    torch_mod.float = "float32"
    torch_mod.int64 = "int64"
    torch_mod.long = "int64"

    # torch.nn — copy attributes into a fresh module so overriding
    # .functional below doesn't mutate browsergrad_grad.nn
    torch_nn = _bg_torch_types.ModuleType("torch.nn")
    for _name in dir(_bg.nn):
        if not _name.startswith("_"):
            setattr(torch_nn, _name, getattr(_bg.nn, _name))

    # torch.nn.functional — copy + PyTorch-name aliases for funcs we
    # name slightly differently internally.
    torch_F = _bg_torch_types.ModuleType("torch.nn.functional")
    for _name in dir(_bg.functional):
        if not _name.startswith("_"):
            setattr(torch_F, _name, getattr(_bg.functional, _name))
    # PyTorch name → browsergrad_grad name
    torch_F.cross_entropy = _bg.functional.cross_entropy_loss
    torch_F.nll = _bg.functional.nll_loss

    torch_nn.functional = torch_F
    torch_mod.nn = torch_nn

    # torch.optim — re-exposed directly (no name differences in v0)
    torch_mod.optim = _bg.optim

    # torch.utils.data — pass through to browsergrad_grad.utils.data
    torch_utils = _bg_torch_types.ModuleType("torch.utils")
    torch_utils.data = _bg.utils.data
    torch_mod.utils = torch_utils

    # Register
    _bg_torch_sys.modules["torch"] = torch_mod
    _bg_torch_sys.modules["torch.nn"] = torch_nn
    _bg_torch_sys.modules["torch.nn.functional"] = torch_F
    _bg_torch_sys.modules["torch.optim"] = _bg.optim
    _bg_torch_sys.modules["torch.utils"] = torch_utils
    _bg_torch_sys.modules["torch.utils.data"] = _bg.utils.data

    return torch_mod
`;
